pkg/resourceutils/scaleDown.go

```diff
@@ -0,0 +1,270 @@
+		if err != nil {
+			return err
+		}
+		if resources != nil {
+			for _, object := range resources.Items {
+				if _, present := resourceMap[ns][cronJobGVK][object.GetName()]; present {
+					err := scaleDownCronJob(dynamicClient, object)
+					if err != nil {
+						return err
+					}
+				}
+			}
+		}
+
+		// 3. CRD Resources
+		for _, applicationRegistration := range crdList.Items {
+			for _, crd := range applicationRegistration.Resources {
+				gvk := crd.GroupVersionKind
+				resources, dynamicClient, err := ListResourcesByGVK(gvk, ns, config)
+				if err != nil {
+					return err
+				}
+				if resources == nil {
+					continue
+				}
+				for _, object := range resources.Items {
+					if _, present := resourceMap[ns][gvk][object.GetName()]; present {
+						err := scaleDownCRDResource(dynamicClient, object, crd, config)
+						if err != nil {
+							return err
+						}
+					}
+				}
+			}
+		}
+	}
+	return nil
+}
+
+func scaleDownApplicationResource(dynamicClient k8sdynamic.ResourceInterface, object unstructured.Unstructured, resourceType metav1.GroupVersionKind) error {
+	content := object.UnstructuredContent()
+	currentReplicas, found, err := unstructured.NestedInt64(content, "spec", "replicas")
+	if err != nil {
+		return err
+	}
+	if !found {
+		currentReplicas = 1
+	}
+	annotations, found, err := unstructured.NestedStringMap(content, "metadata", "annotations")
+	if err != nil {
+		return err
+	}
+	if !found {
+		annotations = make(map[string]string)
+	}
+	migrationReplicasAnnotationValue, replicaAnnotationIsPresent := annotations[migration.StorkMigrationReplicasAnnotation]
+	if currentReplicas == 0 && replicaAnnotationIsPresent {
+		//If the actual replica count is 0 and migrationReplicas annotation is more than 0, then don't override the migrationReplicaAnnotation value.
+		migrationReplicas, err := strconv.ParseInt(migrationReplicasAnnotationValue, 10, 64)
+		if err != nil {
+			return err
+		}
+		if migrationReplicas > 0 {
+			currentReplicas = migrationReplicas
+		}
+	}
+	//we set the actual replica count to 0 to scale down the resource on target cluster
+	err = unstructured.SetNestedField(content, int64(0), "spec", "replicas")
+	if err != nil {
+		return err
+	}
+	annotations[migration.StorkMigrationReplicasAnnotation] = strconv.FormatInt(currentReplicas, 10)
+	err = unstructured.SetNestedStringMap(content, annotations, "metadata", "annotations")
+	if err != nil {
+		return err
+	}
+	opts := &metav1.UpdateOptions{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       resourceType.Kind,
+			APIVersion: resourceType.Group + "/" + resourceType.Version},
+	}
+	_, err = dynamicClient.Update(context.TODO(), &object, *opts, "")
+	if err != nil {
+		return fmt.Errorf("unable to update resource %v %v/%v: %v", strings.ToLower(resourceType.Kind), object.GetNamespace(), object.GetName(), err)
+	}
+	fmt.Printf("successfully updated resource %v %v/%v \n", strings.ToLower(resourceType.Kind), object.GetNamespace(), object.GetName())
```

Comment:
Shouldn't we be using log. in case of Printf?

Reply:
Yes makes sense. Will update everywhere applicable